### PR TITLE
feat(decorators): introduce middleware decorator

### PR DESCRIPTION
**Pull request by kitos**
## Motivation and Context
This low level api that will provide opportunity to create custom middlewares (e.g. for authentication or validation):

### can be used directly

```js
@controller('/foo')
export class FooController implements interfaces.Controller {
  @httpGet('/')
  @middleware((req, res, next) => {
    if (req.user) {
      next()
    } else {
      res.status(401).send('Not authorized.')
    }
  })
  private index() {
    return "private route"
  }
}
```

### or used in a custom decorator:

```js
let authorized = () => middleware((req, res, next) => {
    if (req.user) {
      next()
    } else {
      res.status(401).send('Not authorized.')
    }
  })

@controller('/foo')
export class FooController implements interfaces.Controller {
  @httpGet('/')
  @authorized()
  private index() {
    return "private route"
  }
}
```

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by DusanDimitric** [comment]
This would be so neat to have!

File: src/decorators.ts [hunk 1]
```diff
@@ -1,15 +1,15 @@
-import * as express from "express";
 import { inject, injectable, decorate } from "inversify";
 import { interfaces } from "./interfaces";
 import { TYPE, METADATA_KEY, PARAMETER_TYPE } from "./constants";
+import { getOrCreateMetadata } from "./utils";
 
 export const injectHttpContext = inject(TYPE.HttpContext);
 
-export function controller(path: string, ...middleware: interfaces.Middleware[]) {
+export function controller(path: string, ...middlewares: interfaces.Middleware[]) {
```
Comment (DusanDimitric) [inline src/decorators.ts]: `middleware` is plural, you don't need to add the `s` at the end

File: src/decorators.ts [hunk 2]
```diff
@@ -1,15 +1,15 @@
...
     return function (target: any) {
 
         let currentMetadata: interfaces.ControllerMetadata = {
-            middleware: middleware,
+            middleware: middlewares,
```
Comment (DusanDimitric) [inline src/decorators.ts]: Same here

File: src/decorators.ts [hunk 3]
```diff
@@ -39,54 +39,69 @@ export function controller(path: string, ...middleware: interfaces.Middleware[])
...
 }
 
-export function httpDelete(path: string, ...middleware: interfaces.Middleware[]): interfaces.HandlerDecorator {
-    return httpMethod("delete", path, ...middleware);
+export function httpDelete(path: string, ...middlewares: interfaces.Middleware[]): interfaces.HandlerDecorator {
+    return httpMethod("delete", path, ...middlewares);
 }
 
-export function httpMethod(method: string, path: string, ...middleware: interfaces.Middleware[]): interfaces.HandlerDecorator {
+export function httpMethod(method: string, path: string, ...middlewares: interfaces.Middleware[]): interfaces.HandlerDecorator {
```
Comment (DusanDimitric) [inline src/decorators.ts]: Same for all of these.

File: package.json [hunk 4]
```diff
@@ -57,7 +57,7 @@
     "source-map-support": "0.5.6",
     "supertest": "3.4.2",
     "tslint": "5.9.1",
-    "typescript": "^2.8.3"
+    "typescript": "^3.4.5"
```
Comment (DusanDimitric) [inline package.json]: I would maybe put this version bump in a separate commit.

File: src/decorators.ts [hunk 5]
```diff
@@ -39,54 +39,69 @@ export function controller(path: string, ...middleware: interfaces.Middleware[])
...
+        let controllerMiddlewareMap = getOrCreateMetadata(
+            METADATA_KEY.controllerMethodMiddleware,
+            target.constructor,
+            () => new Map<string, interfaces.Middleware[]>()
+        );
 
-        if (!Reflect.hasMetadata(METADATA_KEY.controllerMethod, target.constructor)) {
-            Reflect.defineMetadata(METADATA_KEY.controllerMethod, metadataList, target.constructor);
+        if (controllerMiddlewareMap.has(methodName)) {
+            controllerMiddlewareMap.get(methodName)!.push(...fns);
```
Comment (DusanDimitric) [inline src/decorators.ts]: This line is kind of confusing, can you unwrap it so it's more understandable?

File: src/decorators.ts (hunk 1 above)
Comment (kitos) [reply src/decorators.ts]: ok, initially I had to do that because of `no-shadow` tslint rule

Comment (DusanDimitric) [reply src/decorators.ts]: Aahh then it's legit I guess :)

File: package.json (hunk 4 above)
Comment (kitos) [reply package.json]: I'll take a look, probably I need this change for some language feature

File: src/decorators.ts (hunk 5 above)
Comment (kitos) [reply src/decorators.ts]: Sure

**Comment by kitos** [comment]
> This would be very nice to have!
> 
> I can immediately see use cases such as:
> 
> ```ts
> @authenticated()
> @hasMinRole('Admin')
> @validatate(ValidationModel)
> ```
> 
> I was looking exactly for this!

Actually I've created this PR to be able to create validation library ;-)

**Comment by DusanDimitric** [comment]
> > This would be very nice to have!
> > I can immediately see use cases such as:
> > ```ts
> > @authenticated()
> > @hasMinRole('Admin')
> > @validatate(ValidationModel)
> > ```
> > 
> > 
> > I was looking exactly for this!
> 
> Actually I've created this PR to be able to create validation library ;-)

That's exactly what is missing for me in `inversify-express-utils`!

**Comment by DusanDimitric** [comment]
Can you please complete this PR so we can escalate the PR and hopefully merge/release?

**Comment by kitos** [comment]
@DusanDimitric I've fixed your comments. So I guess this PR can be merged.
But after it I'll create a couple of more, cause `middleware` decorator is not enough for implementing validation decorator.

**Comment by DusanDimitric** [comment]
Yeah, the CI tests should pass, I don't think it's normal :D

Do the tests pass locally?

**Comment by DusanDimitric** [comment]
@kitos Hey man, how's this PR going?

**Comment by kitos** [comment]
@DusanDimitric sorry I was moving to another country 😆 
I made tests pass locally, but had to update typescript.

And still one of builds (Node.js: stable) doesn't pass.

**Comment by DusanDimitric** [comment]
Hey @kitos, I've been busy for a couple of days as well sorry. Hope your move went well :))

Let's get this over with:

1. How can we collaborate so I can help you with the PR?
2. How do we get it merged since the project seems dead :/

**Comment by DusanDimitric** [comment]
Okay so the tests are passing locally for me but "gulp" is not working..

**Comment by DusanDimitric** [comment]
1. I've cloned your feature branch on my computer
2. Tests pass, everything seems OK
3. `inversify-express-utils` build is failing, this is not your problem. The new node version (12.3.1) does not work with old `gulp`.

We have to contact the maintainers to fix this and then to merge this PR.

**Comment by tspoke** [comment]
Can't wait to get this feature on :)

Will fix https://github.com/inversify/InversifyJS/issues/1121

**Comment by Minibrams** [comment]
Since this PR seems rather dead, I took some inspiration from @kitos and implemented it myself here: #337
Would love this feature in the base package!

**Comment by dcavanagh** [comment]
closed in favor of #337 
Thank you for the workk
